# Wikipedia Notable Life Expectancies

# [Notebook 2 of 4: Data Cleaning](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean_thanak_2022_06_13.ipynb)

## Context

The


## Objective

The

### Data Dictionary

- Feature: Description

## Importing Necessary Libraries

In [1]:
# To structure code automatically
%load_ext nb_black

# To import/export sqlite databases
import sqlite3 as sql

# To help with reading and manipulating data
import pandas as pd
import numpy as np
import re

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)
# To define the maximum number of rows to be displayed in a dataframe
pd.set_option("display.max_rows", 200)

# To supress scientific notations for a dataframe
pd.set_option("display.float_format", lambda x: "%.3f" % x)

# To supress warnings
# import warnings

# warnings.filterwarnings("ignore")

# To set some visualization attributes
pd.set_option("max_colwidth", 150)

<IPython.core.display.Javascript object>

## Data Overview

### Reading, Sampling, and Checking Data Shape

In [2]:
# Reading the wp_life_expect_raw_complete dataset
conn = sql.connect("wp_life_expect_raw_complete.db")
data = pd.read_sql("SELECT * FROM wp_life_expect_raw_complete", conn)

# Making a working copy
df = data.copy()

# Checking the shape
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# Checking first 2 rows of the data
df.head(2)

There are 133900 rows and 6 columns.


month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21  
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12

<IPython.core.display.Javascript object>

In [3]:
# Checking last 2 rows of the data
df.tail(2)

month_year day           name  \
133898  June 2022   9  Oleg Moliboga   
133899  June 2022   9       Zou Jing   

                                                                         info  \
133898    , 69, Russian volleyball player, Olympic champion (1980) and coach.   
133899  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  
133898        https://en.wikipedia.org/wiki/Oleg_Moliboga              2  
133899  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3

<IPython.core.display.Javascript object>

In [4]:
# Checking a sample of the data
df.sample(5)

month_year day                       name  \
1108        June 1994  18                Roger Lebel   
70246       July 2014   8             Maxine Cochran   
4202      August 1995  25             Doug Stegmeyer   
57628  September 2012  13             Edgar Metcalfe   
94594    January 2018  23  Francisco Moreno Martínez   

                                                                                                              info  \
1108                                                                                         , 71, Canadian actor.   
70246  , 87, Canadian politician, Nova Scotia Minister of Transport (1985–1988) and MLA for Lunenburg (1984–1988).   
4202                                                 , 43, American rock bassist and vocalist, suicide by gunshot.   
57628                                                      , 78, English actor, director and writer, liver cancer.   
94594                                                                                       , 86, Spanish cyclist.   

                                                               link  \
1108                      https://en.wikipedia.org/wiki/Roger_Lebel   
70246                  https://en.wikipedia.org/wiki/Maxine_Cochran   
4202                   https://en.wikipedia.org/wiki/Doug_Stegmeyer   
57628                  https://en.wikipedia.org/wiki/Edgar_Metcalfe   
94594  https://en.wikipedia.org/wiki/Francisco_Moreno_Mart%C3%ADnez   

      num_references  
1108               0  
70246              3  
4202               5  
57628             20  
94594              3

<IPython.core.display.Javascript object>

#### Observations:
- There are 133,900 rows and 6 columns.
- `month_year` contains the month and year of death, while `day` contains the day of the month of death.
- `name` is the notable person's name.  It is a nominal feature that will not be used for analysis, but will be maintained for any referencing needs.
- `info` contains multiple items including the notable person's "age, country of citizenship at birth, subsequent country of citizenship (if applicable), reason for notability, (and) cause of death (if known)."
- `link` is the url to the notable person's individual Wikipedia page.  If such a page does not exist, there is either a non-working link (https://en.wikipedia.orgNone), or the link is to a page with a message that the page does not exist for that individual.  `link` is a unique identifier for all entries, except the 6 with the non-working link, which do have unique `name` values from each other.
- `num_references` contains the number of references on the notable person's individual Wikipedia page.  This feature serves as a proxy measure of notability.
- Prior to EDA, our task will be to extract the individual elements that are comined in `month_year` and `info` columns.

### Checking Data Types, Duplicates, and Null Values

In [5]:
# Checking data types and null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133894 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

In [6]:
# Checking duplicate rows
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

In [7]:
# Check percentage of null values by column
df.isnull().sum() / df.count() * 100

month_year       0.000
day              0.000
name             0.004
info             0.000
link             0.000
num_references   0.000
dtype: float64

<IPython.core.display.Javascript object>

In [8]:
# Checking number of missing values per row (not necessary here, but done to keep process standard)
df.isnull().sum(axis=1).value_counts()

0    133894
1         6
dtype: int64

<IPython.core.display.Javascript object>

#### Observations:
- Our dataset was saved to and read from the database without any hiccups.
- As expected, we have 6 entries that are missing `name`, but we will find it in their `info` values.
- All columns are currently of object type.  We will need to appropriately typecast them after separating the information in `month_year` and `info`.

## Data Cleaning

### Addressing Missing `name` Values

In [9]:
# Checking rows with missing name values
missing_name = df[df["name"].isna()]
missing_name

month_year day  name  \
18937      August 2001  11  None   
24985     January 2004  22  None   
27458       March 2005   1  None   
34077        July 2007  11  None   
64769   September 2013  29  None   
106613  September 2019  12  None   

                                                                                      info  \
18937   Kevin Kowalcyk, 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                Vincent Palmer, 37, British criminal.   
27458                                             Barry Stigler, 57, American voice actor.   
34077                                          Nana Gualdi, 75, German singer and actress.   
64769                                 Scott Workman, 47, American stuntman (, , ), cancer.   
106613                                                Thami Shobede, 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- These rows vary from the main set as there is a substring containing the person's name at the start of the `info` string.
- As there are so few rows missing `name`, let us address this issue first.

In [10]:
# For loop to copy name value from info column
treat_rows = missing_name.index
for i in treat_rows:
    info = df.loc[i, "info"]
    info_lst = info.split(sep=",", maxsplit=1)
    df.loc[i, "name"] = info_lst[0]

# Re-check rows
df.loc[treat_rows, :]

month_year day            name  \
18937      August 2001  11  Kevin Kowalcyk   
24985     January 2004  22  Vincent Palmer   
27458       March 2005   1   Barry Stigler   
34077        July 2007  11     Nana Gualdi   
64769   September 2013  29   Scott Workman   
106613  September 2019  12   Thami Shobede   

                                                                                      info  \
18937   Kevin Kowalcyk, 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                Vincent Palmer, 37, British criminal.   
27458                                             Barry Stigler, 57, American voice actor.   
34077                                          Nana Gualdi, 75, German singer and actress.   
64769                                 Scott Workman, 47, American stuntman (, , ), cancer.   
106613                                                Thami Shobede, 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- Missing `name` values have been addressed.
- The names still appear in the `info` value for these rows, but we can address that as we separate the information in that column.

In [11]:
# Re-check info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133900 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

#### Observations:
- We have no remaining missing values.
- Let us treat the `month_year` column next.

### Separating `month` and `year`

In [12]:
# Separating month and year into 2 columns and typecasting year as integer
df.loc[:, "year"] = df["month_year"].apply(lambda x: x.split(sep=" ")[1])
df["year"] = df["year"].astype("int64")

df.loc[:, "month"] = df["month_year"].apply(lambda x: x.split(sep=" ")[0])
df.head(2)

month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

In [13]:
# Dropping month_year column
df.drop("month_year", axis=1, inplace=True)

<IPython.core.display.Javascript object>

### Treating `info`

#### Checking a Sample

In [14]:
# Checking a sample of info
df.sample(5)

day                   name  \
94287   10       William B. Keene   
130789  31  Flemming Quist Møller   
56478    6          Betty Buehler   
108605  21           Deng Hongxun   
23394   14           Robert Stack   

                                                                                           info  \
94287   , 92, American judge (Los Angeles County Superior Court) and television personality ().   
130789                            , 79, Danish animator (, ) and screenwriter (), heart attack.   
56478                                                                   , 90, American actress.   
108605  , 88, Chinese politician and engineer, Communist Party Secretary of Hainan (1990–1993).   
23394                                                 , 84, American film and television actor.   

                                                            link  \
94287             https://en.wikipedia.org/wiki/William_B._Keene   
130789  https://en.wikipedia.org/wiki/Flemming_Quist_M%C3%B8ller   
56478                https://en.wikipedia.org/wiki/Betty_Buehler   
108605                https://en.wikipedia.org/wiki/Deng_Hongxun   
23394                 https://en.wikipedia.org/wiki/Robert_Stack   

       num_references  year     month  
94287               4  2018   January  
130789              8  2022   January  
56478               5  2012      July  
108605              5  2019  December  
23394              29  2003       May

<IPython.core.display.Javascript object>

#### Observations:
- We have leading and trailing characters of comma, period, and white space that can be removed.

#### Removing Leading and Trailing Commas, Whitespace, and Periods

In [15]:
# Removing the leading/trailing commas, periods, and whitespace
df.loc[:, "info"] = df["info"].apply(lambda x: x.strip(" ,."))
df.sample(5)

day                      name  \
26588    3           Janet Backhouse   
63810    4             Betty Babcock   
88611   11             András Kovács   
118112   6  Eugenia Tsoumani-Spentza   
86247   17             Steve Truglia   

                                                                                                                      info  \
26588                                                        66, English manuscripts curator at the British Museum, cancer   
63810   91, American politician, First Lady of Montana (1962–1969), member of Montana House of Representatives (1975–1977)   
88611                                                                                              91, Hungarian filmmaker   
118112                                                                         Greek lawyer and politician, MP (2009–2012)   
86247                                                                                    54, British stuntman (, , ), fall   

                                                                         link  \
26588                           https://en.wikipedia.org/wiki/Janet_Backhouse   
63810                             https://en.wikipedia.org/wiki/Betty_Babcock   
88611   https://en.wikipedia.org/wiki/Andr%C3%A1s_Kov%C3%A1cs_(film_director)   
118112                 https://en.wikipedia.org/wiki/Eugenia_Tsoumani-Spentza   
86247                             https://en.wikipedia.org/wiki/Steve_Truglia   

       num_references  year     month  
26588              21  2004  November  
63810               4  2013    August  
88611               7  2017     March  
118112              1  2020  December  
86247               7  2016  November

<IPython.core.display.Javascript object>

#### Checking the Set of First Substrings (Before First Comma)

In [16]:
# Checking values in the first substring of info (before the first comma)
age_set = set(df["info"].apply(lambda x: x.split(sep=",", maxsplit=1)[0]))
age_set

{'Afghan presidential adviser',
 'French historian and politician',
 'North Korean head of the Unhasu Orchestra',
 '92',
 'British journalist and business executive',
 '98–99',
 'Rwandan journalist',
 'American intelligence officer',
 'c. 66',
 'French doctor and drug test pioneer',
 'Malian actor and comedian',
 '96–97',
 'Indian actor (',
 'Vietnamese politician',
 '40',
 '67–68',
 '42',
 '57-58',
 'American bongo player',
 '79/80',
 'Syrian general',
 'Georgian politician and deputy governor of the Kvemo Kartli region',
 'South African psychotherapist and television production manager (',
 'Burkinabe actor (',
 'common chimpanzee 55',
 'Malawian diplomat involved in leaked diplomatic cable controversy',
 '(岡本喜八)',
 '1–2',
 'Brazilian female pilot',
 '72/73',
 'Sri Lankan lawyer',
 '57–58',
 'Indian Maoist militant',
 'Chilean forester and academic',
 'American librarian and civil rights activist',
 '(Bruno the Bear)',
 '68',
 'American artist',
 'c. 96',
 'French archaeologist',
 'B

<IPython.core.display.Javascript object>

#### Observations:
- Though we can see the age values, there is a lot of other information in the first substring of the `info` value.
- Some entries may be missing the age information.
- Also, age is entered in various formats.
- Let us see how many variations there are.

In [17]:
# Checking number of different values for first substring of info column
len(age_set)

1412

<IPython.core.display.Javascript object>

#### Observations:
- There are 1412 variations for the substring that represents age on Wikipedia.  
- We can start by extracting the age for those entries that begin with the age value for `info`.

#### Extracting `age` from `info` for `info` Values that Begin with "age,"

In [18]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# For loop to extract age from info column and assign to age column, if info starts with 'age,'
for age in ages:
    for i, item in enumerate(df["info"]):
        if item.startswith(age + ","):
            df.loc[i, "age"] = int(age)
df.sample(5)

day             name  \
45561  13     Peter Provan   
32198  23   Ştefan Haukler   
65594  18  Forrest Claunch   
19903  18      Marcel Mule   
73477  23      Norman Wray   

                                                                                                        info  \
45561  73, Australian rugby league footballer, Balmain Tigers premiership captain (1969), after long illness   
32198                                                                            64, Romanian Olympic fencer   
65594                                                             73, American politician, pancreatic cancer   
19903                                                                                100, French saxophonist   
73477                                            91, American Roman Catholic missionary, Alzheimer's disease   

                                                    link num_references  year  \
45561         https://en.wikipedia.org/wiki/Peter_Provan              5  2010   
32198  https://en.wikipedia.org/wiki/%C5%9Etefan_Haukler              3  2006   
65594      https://en.wikipedia.org/wiki/Forrest_Claunch              3  2013   
19903          https://en.wikipedia.org/wiki/Marcel_Mule              1  2001   
73477          https://en.wikipedia.org/wiki/Norman_Wray              9  2014   

          month     age  
45561       May  73.000  
32198  November  64.000  
65594  November  73.000  
19903  December 100.000  
73477  December  91.000

<IPython.core.display.Javascript object>

In [19]:
# Checking for remaining rows missing age
df["age"].isna().sum()

2025

<IPython.core.display.Javascript object>

#### Observations:
- We were able to extract `age` for the vast majority of entries.
- Let us take a look at the remaining entries.

#### Checking Rows with Missing `age`

In [20]:
df[df["age"].isna()]

day                         name  \
268     10           Dominic McGlinchey   
900     20              Fernande Giroux   
1029     6                 Peter Graves   
1895    24  1994 Colombo suicide attack   
2318     1                    Nina Leen   
...     ..                          ...   
133602  24                Khyongla Rato   
133816   4                Mike Omotosho   
133830   4                    Fei Liang   
133837   5                Roman Kutuzov   
133838   5             Stanley Goreraza   

                                                                info  \
268                39–40, Irish republican paramilitary leader, shot   
900                                 Canadian actress and jazz singer   
1029                        English actor and nobleman, heart attack   
1895                                 Notable people killed in the \n   
2318                          Russian-born American photographer for   
...                                                              ...   
133602  98–99, Tibetan Buddhist scholar, founder of The Tibet Center   
133816                                           Nigerian politician   
133830                                  85–86, Chinese herpetologist   
133837                                         Russian major general   
133838                           Zimbabwean military officer, cancer   

                                                                                    link  \
268                                     https://en.wikipedia.org/wiki/Dominic_McGlinchey   
900                                        https://en.wikipedia.org/wiki/Fernande_Giroux   
1029                        https://en.wikipedia.org/wiki/Peter_Graves,_8th_Baron_Graves   
1895    https://en.wikipedia.org/wiki/List_of_attacks_attributed_to_the_LTTE,_1990s#1994   
2318                                             https://en.wikipedia.org/wiki/Nina_Leen   
...                                                                                  ...   
133602                                       https://en.wikipedia.org/wiki/Khyongla_Rato   
133816                                       https://en.wikipedia.org/wiki/Mike_Omotosho   
133830                                           https://en.wikipedia.org/wiki/Fei_Liang   
133837                             https://en.wikipedia.org/wiki/Roman_Kutuzov_(general)   
133838                                    https://en.wikipedia.org/wiki/Stanley_Goreraza   

       num_references  year     month  age  
268               418  1994  February  NaN  
900                 5  1994       May  NaN  
1029                5  1994      June  NaN  
1895               41  1994   October  NaN  
2318                3  1995   January  NaN  
...               ...   ...       ...  ...  
133602             20  2022       May  NaN  
133816             10  2022      June  NaN  
133830              5  2022      June  NaN  
133837             10  2022      June  NaN  
133838              3  2022      June  NaN  

[2025 rows x 8 columns]

<IPython.core.display.Javascript object>

#### Observations:
- We can immediately see 2 apparent patterns.  The first is an age range with a hyphen and the second is age missing altogether.
- Let us do another iteration for the age-range pattern, accepting the lower value as age.

#### Extracting `age` from `info` for `info` Values that Begin with "age-"

In [21]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# For loop to extract age from info column and assign to age column, if info starts with 'age-'
for age in ages:
    for i, item in enumerate(df["info"]):
        if item.startswith(age + "-"):
            df.loc[i, "age"] = int(age)
df.sample(5)

day              name  \
70621   28  Alakbar Mammadov   
116366  27    Barclay Palmer   
35922   27      Irene Stegun   
53612   21     Slavko Ziherl   
49782   16        Chinesinho   

                                                      info  \
70621   84, Azerbaijani Soviet football player and manager   
116366              88, English Olympic shot putter (1956)   
35922                           88, American mathematician   
53612            66, Slovenian psychiatrist and politician   
49782        76, Brazilian footballer, Alzheimer's disease   

                                                  link num_references  year  \
70621   https://en.wikipedia.org/wiki/Alakbar_Mammadov              8  2014   
116366    https://en.wikipedia.org/wiki/Barclay_Palmer              2  2020   
35922       https://en.wikipedia.org/wiki/Irene_Stegun              3  2008   
53612      https://en.wikipedia.org/wiki/Slavko_Ziherl              1  2012   
49782         https://en.wikipedia.org/wiki/Chinesinho              5  2011   

            month    age  
70621        July 84.000  
116366  September 88.000  
35922     January 88.000  
53612     January 66.000  
49782       April 76.000

<IPython.core.display.Javascript object>

In [22]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1841

<IPython.core.display.Javascript object>

#### Observation:
- The last iteration captured nearly 200 values.
- Let us take a look at the remaining rows with missing `name'.

#### Checking Rows with Missing `age`

In [23]:
# Checking rows still missing `age`
df[df["age"].isna()].sample(5)

day                   name  \
11996    3      Ahmet Kayhan Dede   
41319   12          Annesley Dias   
124542  21            Haribhushan   
70914   13            Hani Abbadi   
90317   10  Abu Khattab al-Tunisi   

                                                                            info  \
11996                                                        Turkish Sufi master   
41319                                                        Sri Lankan comedian   
124542                                 Indian politician and guerrilla, COVID-19   
70914   Jordanian politician, member of the House of Representatives (1993–1997)   
90317                                                    Tunisian jihadist, shot   

                                                       link num_references  \
11996       https://en.wikipedia.org/wiki/Ahmet_Kayhan_Dede              0   
41319           https://en.wikipedia.org/wiki/Annesley_Dias              9   
124542            https://en.wikipedia.org/wiki/Haribhushan              6   
70914             https://en.wikipedia.org/wiki/Hani_Abbadi              1   
90317   https://en.wikipedia.org/wiki/Abu_Khattab_al-Tunisi              4   

        year   month  age  
11996   1998  August  NaN  
41319   2009    June  NaN  
124542  2021    June  NaN  
70914   2014  August  NaN  
90317   2017    June  NaN

<IPython.core.display.Javascript object>

#### Observations:
- It almost appears as if our iteration for addressing `info` starting with "age-' missed some values.
- Closer examination reveals that the dash character varies for the remaining values ("-" vs "–").
- So we will iterate again with the larger dash.

In [24]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# For loop to extract age from info column and assign to age column, if info starts with 'age-'
for age in ages:
    for i, item in enumerate(df["info"]):
        if item.startswith(age + "–"):
            df.loc[i, "age"] = int(age)
df.sample(5)

day                 name  \
55084   12  Vladimir Astapovsky   
30048   16      Dennis Kirkland   
48918    6           Gary Moore   
127651  12      Zinaida Korneva   
47932   20          Jim Yardley   

                                                                       info  \
55084             65, Soviet Olympic bronze medal-winning (1976) footballer   
30048   63, British television producer and director, after a short illness   
48918        58, Irish rock guitarist and singer (Thin Lizzy), heart attack   
127651                  99, Russian military veteran and charity fundraiser   
47932                                                 64, English cricketer   

                                                         link num_references  \
55084       https://en.wikipedia.org/wiki/Vladimir_Astapovsky              2   
30048           https://en.wikipedia.org/wiki/Dennis_Kirkland              0   
48918                https://en.wikipedia.org/wiki/Gary_Moore            134   
127651          https://en.wikipedia.org/wiki/Zinaida_Korneva             36   
47932   https://en.wikipedia.org/wiki/Jim_Yardley_(cricketer)              1   

        year     month    age  
55084   2012     April 65.000  
30048   2006  February 63.000  
48918   2011  February 58.000  
127651  2021   October 99.000  
47932   2010  November 64.000

<IPython.core.display.Javascript object>

In [25]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1374

<IPython.core.display.Javascript object>

#### Observation:
- That iteration captured over 400 values.
- Let us take a look at the remaining rows with missing `name'.

#### Checking Rows with Missing `age`

In [26]:
# Checking rows still missing `age`
df[df["age"].isna()].head(100)

day                               name  \
900    20                    Fernande Giroux   
1029    6                       Peter Graves   
1895   24        1994 Colombo suicide attack   
2318    1                          Nina Leen   
2506   22                    Henry Gladstone   
2902   11                 Ernest Kabushemeye   
3296   30                Christopher Chadman   
3359    8                       Carroll Best   
4642   24                Syed Abuzar Bukhari   
5308   12                         Jon Pattis   
5423   29                      Bill Sorensen   
6057    9                       Khaptad Baba   
6078   11        1996 Mount Everest disaster   
6476   17                     TWA Flight 800   
6860   17                        Spiro Agnew   
6998    5                       Elmer Berger   
6999    5                 Wymberley D. Coerr   
7013    7              José Antonio Burciaga   
7094   17                       Bert Hopwood   
7281   13                      Alma Kitchell   
7283   13                         Swami Rama   
7345   21       Syed Mohammed Mukhtar Ashraf   
7400   27                          Jack Penn   
7432    4                         Syd Heylen   
7547   21                      Kell Areskoug   
7930    1                Wilcomb E. Washburn   
8070   20              Ruth Clark (pollster)   
8087   22                       Urmila Bhatt   
8414    5                       Riley Morris   
8421    6                    Barbara Yu Ling   
8464   12                       Moro Lorenzo   
8549   23            Esther Schiff Goldfrank   
8576   26                          Joey Faye   
8719   15                  Saadallah Wannous   
8730   17                     Durgabai Kamat   
8747   21                 Amasa Stone Bishop   
8754   22         Genildo Ferreira de França   
8817   31                  Fazal Haq Mujahid   
8890    9                        Witness Lee   
8918   13                     George Strugar   
8923   15                     Nicholas Danby   
8945   18                      Devon Russell   
8950   19                  Basu Bhattacharya   
9051    4  William Cadogan, 7th Earl Cadogan   
9088    8               Charles P. B. Taylor   
9097    9                     David Pitblado   
9132   14                       Isaac Nicola   
9159   18                    Igor Linchevski   
9257    2                     Antonis Daglis   
9274    4                     Jeanne Calment   
9395   21             Abdul Rahim Ghafoorzai   
9473   31      1997 Pont de l'Alma car crash   
9519    6                    H. W. L. Poonja   
9548   10             Satish Chandra Agarwal   
9704   30                       Nobuo Fujita   
9774    9                    Jean Pasqualini   
9959    3                Satyapramoda Tirtha   
9996    8                   Robert John Kerr   
10063  16                         Padmapriya   
10109  24                            Barbara   
10123  25                 M. Prabhakar Reddy   
10141  28                        Qemal Butka   
10142  28              Wallace H. Clark, Jr.   
10191   4                     G. S. Jayanath   
10339  21                   Sholom Schwadron   
10350  23                       Les Harrison   
10388  27                  Webster Chikabala   
10419  30                         Kim Sun-ja   
10541  13                     Muslim Mulliqi   
10604  21                      Pakoda Kadhar   
10722   5                         Joe Stubbs   
10738   8               Elizabeth Bauer Mock   
10767  11                        Lily Harmon   
10806  17                        Bob Merrill   
10828  19                             Talish   
10910  27                   Henry Munyaradzi   
10934   2                  Maamun al-Kuzbari   
10961   7                 Ziad Rafiq Beydoun   
11054  19                Simon Diedong Dombo   
11077  22                  Shoichi Nishimura   
11121  29                             Loreta   
11232  12                        Xin Fengxia   
11236  12             

<IPython.core.display.Javascript object>

#### Observations:
- The next pattern starts with 'age/', so we will iterate again with it.

In [27]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# For loop to extract age from info column and assign to age column, if info starts with 'age-'
for age in ages:
    for i, item in enumerate(df["info"]):
        if item.startswith(age + "/"):
            df.loc[i, "age"] = int(age)
df.sample(5)

day                 name  \
49341   11          David Brown   
84018   13        Celso Peçanha   
121286   6  Todd R. Klaenhammer   
38960   16       Luisín Landáez   
22220   19         Harry Watson   

                                                                                                         info  \
49341                                                                     69, British cricketer, brain tumour   
84018                                        99, Brazilian politician, Governor of Rio de Janeiro (1961–1962)   
121286                                                         69, American food scientist and microbiologist   
38960                                                                    77, Venezuelan-Chilean cumbia singer   
22220   79, Canadian professional hockey player (Detroit Red Wings, Toronto Maple Leafs, Chicago Black Hawks)   

                                                                      link  \
49341       https://en.wikipedia.org/wiki/David_Brown_(Scottish_cricketer)   
84018                     https://en.wikipedia.org/wiki/Celso_Pe%C3%A7anha   
121286                   https://en.wikipedia.org/wiki/Todd_R._Klaenhammer   
38960               https://en.wikipedia.org/wiki/Luis%C3%ADn_Land%C3%A1ez   
22220   https://en.wikipedia.org/wiki/Harry_Watson_(ice_hockey,_born_1923)   

       num_references  year     month    age  
49341               4  2011     March 69.000  
84018               2  2016      July 99.000  
121286              7  2021     March 69.000  
38960               9  2008  November 77.000  
22220               0  2002  November 79.000

<IPython.core.display.Javascript object>

In [28]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1305

<IPython.core.display.Javascript object>

#### Observation:
- That iteration captured nearly 70 missing values.
- Let us take a look at the remaining rows with missing `name'.

#### Checking Rows with Missing `age`

In [29]:
# Checking rows still missing `age`
df[df["age"].isna()].head(100)

day                               name  \
900    20                    Fernande Giroux   
1029    6                       Peter Graves   
1895   24        1994 Colombo suicide attack   
2318    1                          Nina Leen   
2506   22                    Henry Gladstone   
2902   11                 Ernest Kabushemeye   
3296   30                Christopher Chadman   
3359    8                       Carroll Best   
4642   24                Syed Abuzar Bukhari   
5308   12                         Jon Pattis   
5423   29                      Bill Sorensen   
6057    9                       Khaptad Baba   
6078   11        1996 Mount Everest disaster   
6476   17                     TWA Flight 800   
6860   17                        Spiro Agnew   
6998    5                       Elmer Berger   
6999    5                 Wymberley D. Coerr   
7094   17                       Bert Hopwood   
7281   13                      Alma Kitchell   
7283   13                         Swami Rama   
7345   21       Syed Mohammed Mukhtar Ashraf   
7400   27                          Jack Penn   
7432    4                         Syd Heylen   
7547   21                      Kell Areskoug   
7930    1                Wilcomb E. Washburn   
8070   20              Ruth Clark (pollster)   
8421    6                    Barbara Yu Ling   
8464   12                       Moro Lorenzo   
8576   26                          Joey Faye   
8719   15                  Saadallah Wannous   
8747   21                 Amasa Stone Bishop   
8754   22         Genildo Ferreira de França   
8817   31                  Fazal Haq Mujahid   
8918   13                     George Strugar   
8923   15                     Nicholas Danby   
8945   18                      Devon Russell   
8950   19                  Basu Bhattacharya   
9051    4  William Cadogan, 7th Earl Cadogan   
9088    8               Charles P. B. Taylor   
9097    9                     David Pitblado   
9132   14                       Isaac Nicola   
9159   18                    Igor Linchevski   
9257    2                     Antonis Daglis   
9274    4                     Jeanne Calment   
9395   21             Abdul Rahim Ghafoorzai   
9473   31      1997 Pont de l'Alma car crash   
9519    6                    H. W. L. Poonja   
9548   10             Satish Chandra Agarwal   
9704   30                       Nobuo Fujita   
9774    9                    Jean Pasqualini   
9959    3                Satyapramoda Tirtha   
9996    8                   Robert John Kerr   
10063  16                         Padmapriya   
10109  24                            Barbara   
10123  25                 M. Prabhakar Reddy   
10141  28                        Qemal Butka   
10142  28              Wallace H. Clark, Jr.   
10191   4                     G. S. Jayanath   
10339  21                   Sholom Schwadron   
10350  23                       Les Harrison   
10388  27                  Webster Chikabala   
10419  30                         Kim Sun-ja   
10541  13                     Muslim Mulliqi   
10604  21                      Pakoda Kadhar   
10722   5                         Joe Stubbs   
10738   8               Elizabeth Bauer Mock   
10767  11                        Lily Harmon   
10806  17                        Bob Merrill   
10828  19                             Talish   
10934   2                  Maamun al-Kuzbari   
10961   7                 Ziad Rafiq Beydoun   
11054  19                Simon Diedong Dombo   
11077  22                  Shoichi Nishimura   
11121  29                             Loreta   
11232  12                        Xin Fengxia   
11236  12                   Felicitas Mendez   
11311  23                       Sartaj Singh   
11351  29                  V. Gopalakrishnan   
11379   3                 Venkatesh Kulkarni   
11485  18                   Obaidullah Aleem   
11523  23                      Grace Hartman   
11527  23                   Monroe K. Spears   
11618   5             

<IPython.core.display.Javascript object>

#### Observations:
- The next pattern starts with 'age.', so we will iterate again with it.

In [30]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# For loop to extract age from info column and assign to age column, if info starts with 'age-'
for age in ages:
    for i, item in enumerate(df["info"]):
        if item.startswith(age + "."):
            df.loc[i, "age"] = int(age)
df.sample(5)

day                        name  \
97850    3                 Boris Orlov   
63524   19              A K Azizul Huq   
116840  19                  Tony Lewis   
55582   10            Andreas Shipanga   
47513   16  Aldo Maria Lazzarín Stella   

                                                                                                                 info  \
97850                                                                                    73, Russian gymnastics coach   
63524                                     84,  Bangladeshi civil servant, Comptroller and Auditor General (1983–1989)   
116840                                                      62, English bassist, singer and songwriter (The Outfield)   
55582   80, Namibian politician, Chairman of the Transitional Government of National Unity (1987, 1988), heart attack   
47513                           83, Italian-born Chilean Roman Catholic prelate, Vicar Apostolic of Aysén (1989–1998)   

                                                                 link  \
97850               https://en.wikipedia.org/wiki/Boris_Orlov_(coach)   
63524                    https://en.wikipedia.org/wiki/A_K_Azizul_Huq   
116840            https://en.wikipedia.org/wiki/Tony_Lewis_(musician)   
55582                  https://en.wikipedia.org/wiki/Andreas_Shipanga   
47513   https://en.wikipedia.org/wiki/Aldo_Maria_Lazzar%C3%ADn_Stella   

       num_references  year    month    age  
97850               5  2018     July 73.000  
63524               2  2013     July 84.000  
116840             17  2020  October 62.000  
55582               4  2012      May 80.000  
47513               1  2010  October 83.000

<IPython.core.display.Javascript object>

In [31]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1297

<IPython.core.display.Javascript object>

#### Observation:
- We captured a handful of more missing values with that iteration.
- Let us take a look at the remaining rows with missing `name'.

#### Checking Rows with Missing `age`

In [32]:
# Checking rows still missing `age`
df[df["age"].isna()].head(100)

day                               name  \
900    20                    Fernande Giroux   
1029    6                       Peter Graves   
1895   24        1994 Colombo suicide attack   
2318    1                          Nina Leen   
2506   22                    Henry Gladstone   
2902   11                 Ernest Kabushemeye   
3296   30                Christopher Chadman   
3359    8                       Carroll Best   
4642   24                Syed Abuzar Bukhari   
5308   12                         Jon Pattis   
5423   29                      Bill Sorensen   
6057    9                       Khaptad Baba   
6078   11        1996 Mount Everest disaster   
6476   17                     TWA Flight 800   
6860   17                        Spiro Agnew   
6998    5                       Elmer Berger   
6999    5                 Wymberley D. Coerr   
7094   17                       Bert Hopwood   
7281   13                      Alma Kitchell   
7283   13                         Swami Rama   
7345   21       Syed Mohammed Mukhtar Ashraf   
7400   27                          Jack Penn   
7432    4                         Syd Heylen   
7547   21                      Kell Areskoug   
7930    1                Wilcomb E. Washburn   
8070   20              Ruth Clark (pollster)   
8421    6                    Barbara Yu Ling   
8464   12                       Moro Lorenzo   
8576   26                          Joey Faye   
8719   15                  Saadallah Wannous   
8747   21                 Amasa Stone Bishop   
8754   22         Genildo Ferreira de França   
8817   31                  Fazal Haq Mujahid   
8923   15                     Nicholas Danby   
8945   18                      Devon Russell   
8950   19                  Basu Bhattacharya   
9051    4  William Cadogan, 7th Earl Cadogan   
9088    8               Charles P. B. Taylor   
9097    9                     David Pitblado   
9132   14                       Isaac Nicola   
9159   18                    Igor Linchevski   
9257    2                     Antonis Daglis   
9274    4                     Jeanne Calment   
9395   21             Abdul Rahim Ghafoorzai   
9473   31      1997 Pont de l'Alma car crash   
9519    6                    H. W. L. Poonja   
9548   10             Satish Chandra Agarwal   
9704   30                       Nobuo Fujita   
9774    9                    Jean Pasqualini   
9959    3                Satyapramoda Tirtha   
9996    8                   Robert John Kerr   
10063  16                         Padmapriya   
10109  24                            Barbara   
10123  25                 M. Prabhakar Reddy   
10141  28                        Qemal Butka   
10142  28              Wallace H. Clark, Jr.   
10191   4                     G. S. Jayanath   
10339  21                   Sholom Schwadron   
10350  23                       Les Harrison   
10388  27                  Webster Chikabala   
10419  30                         Kim Sun-ja   
10541  13                     Muslim Mulliqi   
10604  21                      Pakoda Kadhar   
10722   5                         Joe Stubbs   
10738   8               Elizabeth Bauer Mock   
10767  11                        Lily Harmon   
10806  17                        Bob Merrill   
10828  19                             Talish   
10934   2                  Maamun al-Kuzbari   
10961   7                 Ziad Rafiq Beydoun   
11054  19                Simon Diedong Dombo   
11077  22                  Shoichi Nishimura   
11121  29                             Loreta   
11232  12                        Xin Fengxia   
11236  12                   Felicitas Mendez   
11311  23                       Sartaj Singh   
11351  29                  V. Gopalakrishnan   
11379   3                 Venkatesh Kulkarni   
11485  18                   Obaidullah Aleem   
11523  23                      Grace Hartman   
11527  23                   Monroe K. Spears   
11618   5                         B. M. Shah   
11624   6             

<IPython.core.display.Javascript object>

#### Observations:
- There is at least one entry that escaped the net of that last iteration.  It can be hard-coded later.
- The next pattern starts with "age ", so will will iterate again with it.

In [33]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# For loop to extract age from info column and assign to age column, if info starts with 'age-'
for age in ages:
    for i, item in enumerate(df["info"]):
        if item.startswith(age + " "):
            df.loc[i, "age"] = int(age)
df.sample(5)

day             name                                         info  \
88898   25    Louis Feldman               90, American classical scholar   
111236  29    Chandan Singh                  94, Indian military officer   
56680   18    Pancho Martin           86, Cuban racehorse trainer (Sham)   
76216   11   Frankie Sodano            84, American Olympic boxer (1948)   
38670   20  Sœur Emmanuelle  99, Belgian-born French nun, natural causes   

                                                                  link  \
88898                      https://en.wikipedia.org/wiki/Louis_Feldman   
111236  https://en.wikipedia.org/wiki/Chandan_Singh_(Air_Vice_Marshal)   
56680                      https://en.wikipedia.org/wiki/Pancho_Martin   
76216                     https://en.wikipedia.org/wiki/Frankie_Sodano   
38670               https://en.wikipedia.org/wiki/S%C5%93ur_Emmanuelle   

       num_references  year    month    age  
88898               5  2017    March 90.000  
111236             13  2020    March 94.000  
56680              10  2012     July 86.000  
76216               1  2015      May 84.000  
38670               6  2008  October 99.000

<IPython.core.display.Javascript object>

In [34]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1271

<IPython.core.display.Javascript object>

#### Observations:
- That iteration captured 26 more missing values.
- Let us look again at the remaining rows.

#### Checking Rows with Missing `age`

In [35]:
# Checking rows still missing `age`
df[df["age"].isna()].head(100)

day                               name  \
900    20                    Fernande Giroux   
1029    6                       Peter Graves   
1895   24        1994 Colombo suicide attack   
2318    1                          Nina Leen   
2506   22                    Henry Gladstone   
2902   11                 Ernest Kabushemeye   
3296   30                Christopher Chadman   
3359    8                       Carroll Best   
4642   24                Syed Abuzar Bukhari   
5308   12                         Jon Pattis   
5423   29                      Bill Sorensen   
6057    9                       Khaptad Baba   
6078   11        1996 Mount Everest disaster   
6476   17                     TWA Flight 800   
6860   17                        Spiro Agnew   
6998    5                       Elmer Berger   
6999    5                 Wymberley D. Coerr   
7094   17                       Bert Hopwood   
7281   13                      Alma Kitchell   
7283   13                         Swami Rama   
7345   21       Syed Mohammed Mukhtar Ashraf   
7400   27                          Jack Penn   
7432    4                         Syd Heylen   
7930    1                Wilcomb E. Washburn   
8070   20              Ruth Clark (pollster)   
8421    6                    Barbara Yu Ling   
8464   12                       Moro Lorenzo   
8576   26                          Joey Faye   
8719   15                  Saadallah Wannous   
8747   21                 Amasa Stone Bishop   
8754   22         Genildo Ferreira de França   
8817   31                  Fazal Haq Mujahid   
8923   15                     Nicholas Danby   
8945   18                      Devon Russell   
8950   19                  Basu Bhattacharya   
9051    4  William Cadogan, 7th Earl Cadogan   
9088    8               Charles P. B. Taylor   
9097    9                     David Pitblado   
9132   14                       Isaac Nicola   
9159   18                    Igor Linchevski   
9257    2                     Antonis Daglis   
9395   21             Abdul Rahim Ghafoorzai   
9473   31      1997 Pont de l'Alma car crash   
9519    6                    H. W. L. Poonja   
9548   10             Satish Chandra Agarwal   
9704   30                       Nobuo Fujita   
9774    9                    Jean Pasqualini   
9959    3                Satyapramoda Tirtha   
9996    8                   Robert John Kerr   
10063  16                         Padmapriya   
10109  24                            Barbara   
10123  25                 M. Prabhakar Reddy   
10141  28                        Qemal Butka   
10142  28              Wallace H. Clark, Jr.   
10191   4                     G. S. Jayanath   
10339  21                   Sholom Schwadron   
10350  23                       Les Harrison   
10388  27                  Webster Chikabala   
10419  30                         Kim Sun-ja   
10541  13                     Muslim Mulliqi   
10604  21                      Pakoda Kadhar   
10722   5                         Joe Stubbs   
10738   8               Elizabeth Bauer Mock   
10767  11                        Lily Harmon   
10806  17                        Bob Merrill   
10828  19                             Talish   
10934   2                  Maamun al-Kuzbari   
10961   7                 Ziad Rafiq Beydoun   
11054  19                Simon Diedong Dombo   
11077  22                  Shoichi Nishimura   
11121  29                             Loreta   
11232  12                        Xin Fengxia   
11236  12                   Felicitas Mendez   
11311  23                       Sartaj Singh   
11351  29                  V. Gopalakrishnan   
11379   3                 Venkatesh Kulkarni   
11485  18                   Obaidullah Aleem   
11523  23                      Grace Hartman   
11527  23                   Monroe K. Spears   
11618   5                         B. M. Shah   
11624   6                        Jatoe Kaleo   
11717  18                     Nazim Panipati   
11734  20             

<IPython.core.display.Javascript object>

#### Observations:
- This sample shows entries that are missing age information, or the age is imbedded in the middle of the string. 
- Let us remove rows without any age information, as they will not add to the analysis.

In [36]:
# Creating a list of ages
ages = pd.Series(np.arange(1, 150)).astype("string")

# List of index of entries to drop due to no age information
drop_list = []

# For loop to add index of entries to drop_list if no age information present
for i in df[df["age"].isna()].index:
    if not any(age in df.loc[i, "info"] for age in ages):
        drop_list.append(i)
df.loc[drop_list, :].sample(50)

day                                name  \
75011   10                     Mohammed Kaliel   
15639    3                    Ernesto Layaguin   
46870   24               Bulle Hassan Mo'allim   
16208   10                   Denis O'Conor Don   
8421     6                     Barbara Yu Ling   
110073  17                      Pandhari Juker   
66946   26                           Tom Nyuma   
121810  23                    Gilles Rossignol   
27407   23                        José Cruxent   
65757   26                        Himachal Som   
9473    31       1997 Pont de l'Alma car crash   
45214   18  Abu Abdullah al-Rashid al-Baghdadi   
126420  27                    Noor Islam Dawar   
29146   20         Saadoun Sughaiyer al-Janabi   
27013    4                       Ali Al-Haidri   
73106    5                         Yashaw Adem   
31863   11                    Raad Mutar Saleh   
125803   6                               Barry   
46269    7                          Bob Harvie   
114842  28                    Mohamed Mashally   
57213   19                      Ghazi al-Sadiq   
17014   24             Mohammad Aslam Watanjar   
59675    2                         Joe McGrath   
37056   13                 Maheswary Velautham   
109314  19                       Allah Thérèse   
15936   23       Mishari bin Abdulaziz Al Saud   
13978    1                     Yaakov Weinberg   
73272   14        Fatima bint Rashid Al Nuaimi   
123078  30                        Jalal Yousef   
15979    1                       Mary Mdziniso   
54304   28                        Ian Gallahar   
13707    9                       Wynona Lipman   
109332  20                Gyanendra Nath Pande   
30431   14                        Henry Callow   
14016    8                              Chaino   
25508   26                      Scott Williams   
132432  31                  Easa Saleh Al Gurg   
50989   17                   Jan Mohammad Khan   
26907   21                     Michael Forrest   
24295    8                      David Margolis   
56212   18                     Salem Ali Qatan   
77918   18              Abu Muslim al-Turkmani   
50099    8                  Huthaifa al-Batawi   
127531   8                      Marina Galanou   
19686   19         Pul-i-Estikam bridge ambush   
14064   15                  Kamalendumati Shah   
12169    4               Hernando Durán Dussán   
22230   21                Arturo Guzman Decena   
21448   25                 Abdur Rahman Badawi   
27240    3                         Raul Usupov   

                                                                                                                info  \
75011                                                                           Nigerian army officer and politician   
15639               hospital corpsman of the Philippine Marine Corps and posthumous recipient of  the Medal of Valor   
46870                  Somali politician, member of the Transitional Federal Parliament, victim of Muna Hotel attack   
16208                                                              O'Conor Don, hereditary Chief of the Name O'Conor   
8421                                                                                       Singapore-British actress   
110073                                                                                         Indian make-up artist   
66946                                                                 Sierra Leonean politician and military officer   
121810                                                                                      French writer and editor   
27407                                                                                       Venezuelan archaeologist   
65757                                                                                                Indian diplomat   
9473                                                                                        People killed in the :\n   
45214           

<IPython.core.display.Javascript object>

In [37]:
# Checking the number of entries that will be dropped
print(len(drop_list))

# Dropping the rows without age information
df.drop(drop_list, axis=0, inplace=True)

925


<IPython.core.display.Javascript object>

In [39]:
# Checking df shape after dropping rows
df.shape

(132975, 8)

<IPython.core.display.Javascript object>

#### Observations:
- We have successfully dropped the entries lacking age data.
- Let us examine the remaining rows missing `age`.

#### Checking Rows with Missing `age`

In [44]:
# Checking rows still missing `age`
df[df["age"].isna()].sample(100)

day                          name  \
34411   21                     Čabulītis   
77501   22                 Martin Storey   
109765   5               Mohammad Shafiq   
30593    6             Flight Lieutenant   
45575   15             Gabriel Bien-Aimé   
131258  14          Khayal Zaman Orakzai   
59193   10                   Harry Iauko   
36946    2                    Justin Yak   
90591   25     Agha Shahbaz Khan Durrani   
82352   11               A. R. Surendran   
119176   7            Munira Yamin Satti   
34162   22                  John Burnett   
114399  12            Mohamed Abdi Hashi   
82565   22                   Soran Singh   
127556   9                      Ali Atwa   
31859   11               Benito Martínez   
27352   16                Narriman Sadek   
129470  20             Mamadú Iaia Djaló   
30748   26            Mahmoud al-Majzoub   
71714   23                  Mullah Ghani   
99086    9              Simon Adut Yuang   
13030   22                 Graciela Quan   
111072  25               Robert Levinson   
16068   15                           NRG   
30571    2              Sam Mokuahi, Jr.   
71089   21                  Raed al Atar   
93890   23                    James Ligo   
127168  24           Ishfaq Nadeem Ahmad   
128020  27               Gerald R. Young   
128764  25                 Galal Ibrahim   
123722  20          Tshoganetso Tongwane   
40795   22               Dudley Thornton   
6860    17                   Spiro Agnew   
26681   18                 Alfred Maseng   
129375  17           Dah Sagbadjou Glele   
57928    1                 Chronox Manek   
47881   15                Nimr al-Khatib   
30907   17      Cláudio Besserman Vianna   
109656   1             Admiral K. Sangma   
112797  12               Henriette Conté   
98051   14                  Mick Langley   
87934    8            Mohamud Muse Hersi   
30980   26                      Bear JJ1   
30955   23                       Harriet   
122785  21                Erasmo Vásquez   
119529  16  Christian Myekeni Ntshangase   
44313   12            Saleban Olad Roble   
25232    8               Muhammad Zaidan   
25794   18                  Paul Johnson   
116330  25              Simos Simopoulos   
75398   28                 Tuti Yusupova   
38472    1                Arlene Sherman   
56312   24               Lonesome George   
113690  14              Tawfiq al-Yasiri   
130238  15              Claire D. Clarke   
132667  10            Desmond Brathwaite   
114582  18                  Martha Mmola   
126572   1                Margaret Stone   
67603   24              Hamid Nawaz Khan   
27848   22                     Mary Dann   
86118   11            Jum Jainudin Akbar   
117796  24            Hussein Al-Zuhairi   
30475   20                Stanley Hiller   
87568   23                   Kang Ki-sop   
6057     9                  Khaptad Baba   
111847  12            Khalif Mumin Tohow   
68137   22            Hamza Abu al-Haija   
119360  12           Lingson Belekanyama   
118150   7                       Tasiman   
62710    4                   Luo Meizhen   
42596    5                         Evile   
108081  26                  Gulzar Ahmed   
27398   22                    Lee Eun-ju   
40227    2                 Halit Balamir   
116170  18                  A. Kalanithi   
119811  23              Makhosi Vilakati   
29357   15              Agenore Incrocci   
34583    9            Sir Tasker Watkins   
68130   21                  Colin Turner   
87593   24                Lualemaga Faoa   
30849    8                  Jack Jackson   
31533    1               Kyffin Williams   
118621  23              Loyiso Mpumlwana   
114584  18           David Romero Ellner   
18937   11                Kevin Kowalcyk   
108699  25                      Niju Ram   
113818  19      Thandi Mpambo-Sibhukwana   
131765   3                  William Samb   
31576    6  Mohammed Taha Mohammed Ahmed   
64769   29                 Scott Workman   
497

<IPython.core.display.Javascript object>